core/trino-main/src/main/java/io/trino/execution/CreateFunctionTask.java

```diff
@@ -0,0 +1,172 @@
+import static io.trino.sql.analyzer.SemanticExceptions.semanticException;
+import static io.trino.sql.routine.SqlRoutineAnalyzer.isRunAsInvoker;
+import static java.util.Objects.requireNonNull;
+
+public class CreateFunctionTask
+        implements DataDefinitionTask<CreateFunction>
+{
+    private final Optional<CatalogSchemaName> functionSchema;
+    private final SqlParser sqlParser;
+    private final Metadata metadata;
+    private final FunctionManager functionManager;
+    private final AccessControl accessControl;
+    private final LanguageFunctionManager languageFunctionManager;
+
+    @Inject
+    public CreateFunctionTask(
+            SqlEnvironmentConfig sqlEnvironmentConfig,
+            SqlParser sqlParser,
+            Metadata metadata,
+            FunctionManager functionManager,
+            AccessControl accessControl,
+            LanguageFunctionManager languageFunctionManager)
+    {
+        this.functionSchema = functionSchema(sqlEnvironmentConfig);
+        this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.functionManager = requireNonNull(functionManager, "functionManager is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+        this.languageFunctionManager = requireNonNull(languageFunctionManager, "languageFunctionManager is null");
+    }
+
+    @Override
+    public String getName()
+    {
+        return "CREATE FUNCTION";
+    }
+
+    @Override
+    public ListenableFuture<Void> execute(CreateFunction statement, QueryStateMachine stateMachine, List<Expression> parameters, WarningCollector warningCollector)
+    {
+        Session session = stateMachine.getSession();
+
+        FunctionSpecification function = statement.getSpecification();
+        QualifiedObjectName name = qualifiedFunctionName(functionSchema, statement, function.getName());
+
+        accessControl.checkCanCreateFunction(session.toSecurityContext(), name);
+
+        String formatted = formatSql(function);
+        verifyFormattedFunction(formatted, function);
+
+        languageFunctionManager.verifyForCreate(session, formatted, functionManager, accessControl);
+
+        String signatureToken = languageFunctionManager.getSignatureToken(function.getParameters());
+
+        // system path elements currently are not stored
+        List<CatalogSchemaName> path = session.getPath().getPath().stream()
+                .filter(element -> !element.getCatalogName().equals(GlobalSystemConnector.NAME))
+                .toList();
+
+        Optional<String> owner = isRunAsInvoker(function) ? Optional.empty() : Optional.of(session.getUser());
+
+        LanguageFunction languageFunction = new LanguageFunction(signatureToken, formatted, path, owner);
+
+        boolean replace = false;
+        if (metadata.languageFunctionExists(session, name, signatureToken)) {
```

Comment:
In the future we should likely use `Metadata.getFunctions(Session session, CatalogSchemaFunctionName name)`. So you don't try to create a non-language function that already exists.  I'm not concerned about this right now as I don't expect a catalog to support both anytime soon.